src/pubsub/ua_pubsub_reader.c

```diff
@@ -1839,4 +1722,270 @@ UA_PubSubConnection_processNetworkMessage(UA_Server *server, UA_PubSubConnection
+#ifdef UA_ENABLE_PUBSUB_BUFMALLOC
+        useNormalAlloc();
+#endif /* UA_ENABLE_PUBSUB_BUFMALLOC */
+        return UA_STATUSCODE_UNCERTAIN;
+    }
+
+    /* Check the decoded message is the expected one
+     * TODO: PublisherID check after modification in NM to support all datatypes */
+    if((dataSetReader->bufferedMessage.nm->groupHeader.writerGroupId !=
+        dataSetReader->config.writerGroupId) ||
+       (*dataSetReader->bufferedMessage.nm->payloadHeader.dataSetPayloadHeader
+           .dataSetWriterIds != dataSetReader->config.dataSetWriterId)) {
+        UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                    "PubSub receive. Unknown message received. Will not be processed.");
+        UA_DataSetMessage_freeDecodedPayload(
+            dataSetReader->bufferedMessage.nm->payload.dataSetPayload.dataSetMessages);
+#ifdef UA_ENABLE_PUBSUB_BUFMALLOC
+        useNormalAlloc();
+#endif /* UA_ENABLE_PUBSUB_BUFMALLOC */
+        return UA_STATUSCODE_UNCERTAIN;
+    }
+
+    UA_DataSetReader_process(
+        server, dataSetReader,
+        dataSetReader->bufferedMessage.nm->payload.dataSetPayload.dataSetMessages);
+
+    UA_DataSetMessage_freeDecodedPayload(
+        dataSetReader->bufferedMessage.nm->payload.dataSetPayload.dataSetMessages);
+
+#ifdef UA_ENABLE_PUBSUB_BUFMALLOC
+    useNormalAlloc();
+#endif /* UA_ENABLE_PUBSUB_BUFMALLOC */
+
+    /* Minimum ethernet packet size is 64 bytes where the header size is 14
+     * bytes and FCS size is 4 bytes so remaining minimum payload size of
+     * ethernet packet is 46 bytes */
+    /* TODO: Need to handle padding bytes for UDP */
+    if(((*currentPosition - previousPosition) < MIN_PAYLOAD_SIZE_ETHERNET) &&
+       (strncmp((const char *)connection->config->transportProfileUri.data,
+                "http://opcfoundation.org/UA-Profile/Transport/pubsub-eth-uadp",
+                (size_t)(connection->config->transportProfileUri.length)) == 0)) {
+        paddingBytes =
+            (MIN_PAYLOAD_SIZE_ETHERNET - ((*currentPosition) - previousPosition));
+        (*currentPosition) += paddingBytes; /* During multiple receive, move the position
+                                               to handle padding bytes */
+    }
+    return UA_STATUSCODE_GOOD;
+}
+
+UA_StatusCode
+receiveBufferedNetworkMessage(UA_Server *server, UA_ReaderGroup *readerGroup,
+                              UA_PubSubConnection *connection) {
+    UA_ByteString buffer;
+    buffer.length = RECEIVE_MSG_BUFFER_SIZE;
+    buffer.data = ReceiveMsgBuffer;
+
+    UA_StatusCode rv = connection->channel->receive(
+        connection->channel, &buffer, NULL, readerGroup->config.timeout);
+
+    // TODO attention: here rv is ok if UA_STATUSCODE_GOOD != rv
+    UA_CHECK_WARN(!UA_StatusCode_isBad(rv), return rv,
+                  &server->config.logger, UA_LOGCATEGORY_SERVER,
+                  "SubscribeCallback(): Connection receive failed!");
+    if(buffer.length > 0) {
```

Comment:
You can move the if(rt) inside the while-loop and reuse the same while-loop for both cases.
That also lets you remove the outer if. Just do a while() instead of do {} while.

Reply:
This would mean an additional "if" statement inside a while loop.
I wanted to avoid that, although surely it might not be of any relevance.

otherwise I had a solution once which included a function pointer, do that instead?:

```c
    UA_StatusCode (*decodeAndProcessNetworkMessageFun)(UA_Server *server, UA_ReaderGroup *readerGroup,
                                     UA_PubSubConnection *connection, size_t previousPosition,
                                     UA_ByteString *buffer, size_t *currentPosition) = NULL;

    if (readerGroup->config.rtLevel == UA_PUBSUB_RT_FIXED_SIZE) {
        decodeAndProcessNetworkMessageFun = decodeAndProcessNetworkMessageRT;
    } else {
        decodeAndProcessNetworkMessageFun = decodeAndProcessNetworkMessage;
    }

    if(buffer.length > 0) {
        size_t currentPosition = 0;
        size_t previousPosition = 0;
        do {
            rv = decodeAndProcessNetworkMessageFun(server, readerGroup, connection,
                                                      previousPosition, &buffer,
                                                      &currentPosition);
            UA_CHECK_STATUS_WARN(
                rv, (void) 0, &server->config.logger, UA_LOGCATEGORY_SERVER,
                "SubscribeCallback(): receive message failed");

                previousPosition = currentPosition;

        } while((buffer.length) > currentPosition);
    }
```